### DISCUSSION TITLE
max-batch-size对显存占用的影响

### DISCUSSION THREAD

**Discussion by idontlikelongname** [InternLM/lmdeploy#2479] [Q&A]
使用W4A16量化后的Meta-Llama-3.1-70B-Instruct，在H800-80G测试模型吞吐量，使用脚本benchmark/profile_throughput.py。concurrency分别等于16和128时，GPU显存占用并没有明显的区别。
请问为什么会产生这种现象，我理解显存占用量与bath size的大小近似线性增长，从实际表现上看好像并不是这样。
请问Autoregression类模型的显存占用和batch size之间是什么关系？

**Comment by lvhan028** [comment]
不是线性增长的关系，和 batch size 无关。
可以看下 https://lmdeploy.readthedocs.io/en/latest/get_started/get_started.html 中 Note 部分关于内存分配策略的说明。

**Reply by idontlikelongname** [reply]
谢谢回复，提问时我陷入了传统CNN的定式思维，即显存占用的大小与输入batch之间存在类似线性增长的关系。
但是GPT类的LLM模型，使用transformer结构，推理时使用AutoRegression方式得到所有输出，显存占用的分配关系与传统CNN似乎有一些不同。
AR模型显存占用分两部分
1. 模型权重：模型参数占用的内存，例如7B的模型，以fp16精度加载时，显存占用大小为14GB
2. KV缓存：内存中缓存的注意力张量
在批处理时，不会影响模型权重的显存占用量，KV缓存的数量会成比例增加。但是KV缓存相比显存占用的比例依旧非常小。
以下公式描述了适用于当今大多数常用LLM架构的KV缓存大小。

每个token的KV缓存大小=2 * (num_layers) * (num_heads * dim_head) * precision_in_bytes

例如对于16位精度的LLama2-7B模型，如果batch=16，则KV缓存的大小为128x4096x2x32x4096x2，即32GB。
如果使用LMDeploy的KV-INT4，则缓存占用缩小到8GB


参考文章
https://developer.nvidia.com/zh-cn/blog/mastering-llm-techniques-inference-optimization/

**Reply by idontlikelongname** [reply]
请问我上述的理解是否正确，同时我注意到LMDeploy的配置中，KV缓存的长度似乎被设置成64？

**Reply by lvhan028** [reply]
“例如对于16位精度的LLama2-7B模型，如果batch=16，则KV缓存的大小为128x4096x2x32x4096x2，即32GB。”
计算公式和batch无关
没看懂这个大小是怎么来的。
num_heads 表示 kv head 的数量，dim_head表示 kv head 的dim，目前 turbomind engine只支持了 128

“KV缓存的长度似乎被设置成64”
是的。一个 KV block 对应的是 `cache_block_seq_len` 个 token的 kv 缓存，它的默认值是 64

**Reply by idontlikelongname** [reply]
batch size等于128
token长度等于4096
num_layers等于32
num_heads * dim_head 等于 hidden_size，4096
precision_in_bytes，fp16等于2
这是我对这几个参数的理解。

**Reply by lvhan028** [reply]
again, 和batch size 无关。和要存的token数量有关

**Reply by idontlikelongname** [reply]
抱歉再次打扰。
LMDeploy有预留显存的机制，通过cache-max-entry-count设置预留显存的大小，所以实际显存占用量与batch size没有关联。
可是如果cache-max-entry-count，KV cache的显存占用与batch size之间是什么关系呢？
目前实际测试下来的结果如下：
模型参数和配置：LLama-3.1-70B-INT4模型，quant-policy=4，batch size每增加1，显存占用增加1MB（比较粗略的估算）
但是上述数据增长的比例，我无法用一个准确的数学关系将显存增长与模型参数关联起来。

请问能提供一些关于问题的思路吗？
